### PR TITLE
Fix for issues/610 and issues/643

### PR DESCRIPTION
**Pull request by karthick-vinod**
Fix for https://github.com/gregnb/mui-datatables/issues/610 and https://github.com/gregnb/mui-datatables/issues/643

### CODE REVIEW

**Comment by gabrielliwerant** [comment]
Just a heads up, I'm not ignoring this PR. I have in fact begun to review it, and will hopefully have feedback for you within the next few days.

**Comment by gabrielliwerant** [comment]
@karthick-vinod Firstly, thank you for being proactive and attempting to fix this issue. It is an important issue, however, I've been working on it a bit and thinking about it a lot, and I think there are fundamental problems causing it in the library at the moment.

We are extending the use of the `componentDidMount` lifecycle here as a way to refresh derived state. Attempting to use lifecycle hooks in this way is recommended against by react as it causes numerous issues (and also turns out to be very difficult to test): https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html.

This is not your fault, as you're just following the existing conventions. However, we do indeed have edge cases with these kinds of fixes, and in this case, it does revert prior fixes/issues (https://github.com/gregnb/mui-datatables/issues/305, https://github.com/gregnb/mui-datatables/pull/585, https://github.com/gregnb/mui-datatables/pull/597).

Fundamentally, the library is attempting to allow users to have it both ways when it comes to state. It is attempting to allow changes via props to be reflected in the table AND it is attempting to manage an internal derived state which should also update on user actions and prop updates. This entails two separate sources of truth: the props sent to the table, and the internal state of the table as derived from those props. I think this project is too complex to allow both approaches without becoming a maintenance nightmare.

I think the library should shift to forcing most changes to occur via props instead of derived internal state, but this will be a major shift in the API and cause a lot of breaking changes. What I'd like to do at the moment is use this as an opportunity to shift the library in that direction, without finishing it, hopefully fixing the bugs here without breaking anything else and paving the way forward. I will invite you to take a look when I'm done.

If it seems like it is taking too long for the changes, I will return to this PR and see if there is a way it can be cleaned up to prevent the regressions so that we are not without a fix for too long.

**Comment by gabrielliwerant** [comment]
@karthick-vinod I'll go back to reviewing this PR and see if we can get a fix in here that doesn't include regressions. Work on the refactor will take place separately, as it is too large to handle in response to critical bugs: https://github.com/gregnb/mui-datatables/issues/679.

**Comment by tiagofernandez** [comment]
It'd be nice to get this fixed in v2, as I think many of us are stuck with v2.0.0 – thanks @gabrielliwerant for looking into this!

File: src/MUIDataTable.js [hunk 1]
```diff
@@ -196,8 +196,16 @@ class MUIDataTable extends React.Component {
...
   componentDidUpdate(prevProps) {
-    if (this.props.data !== prevProps.data || this.props.columns !== prevProps.columns) {
-      this.setTableData(this.props, TABLE_LOAD.INITIAL, () => {
+    if (
+      this.props.data !== prevProps.data ||
+      this.props.columns !== prevProps.columns ||
+      this.props.options !== prevProps.options
+    ) {
+      if (this.props.options !== prevProps.options) {
+        this.getDefaultOptions(this.props);
```
Comment (gabrielliwerant) [inline src/MUIDataTable.js]: Reverting back to the default props here is problematic as it conflicts with changes in state. For example, this opens a regression where setting a page number in props and then changing the page number in the table causes any change in options to revert back to the original page.

File: src/MUIDataTable.js [hunk 2]
```diff
@@ -196,8 +196,16 @@ class MUIDataTable extends React.Component {
...
+        this.setTableOptions(this.props);
+      }
+      this.setTableData(prevProps, TABLE_LOAD.INITIAL, () => {
```
Comment (gabrielliwerant) [inline src/MUIDataTable.js]: Using `prevProps` here opens this regression: https://github.com/gregnb/mui-datatables/pull/597.

File: test/MUIDataTable.test.js [hunk 3]
```diff
@@ -211,6 +220,67 @@ describe('<MUIDataTable />', function() {
     assert.deepEqual(state.data, expectedResult);
   });
 
+  it('should correctly re-build internal table data and displayData structure with columns prop change', () => {
```
Comment (gabrielliwerant) [inline test/MUIDataTable.test.js]: Much appreciate the attempt to test the library and the changes, however, both of these tests pass in `master` or with all your other changes here removed, meaning that they don't actually demonstrate that your code fixes the problem. This is part of the difficulty with logic in the lifecycle methods, in that it is very difficult to mimic their behavior in the same way as in actual usage of the library.

**Comment by patorjk** [comment]
Should be fixed as of version 2.15.0
